mne/annotations.py

```diff
@@ -517,3 +519,82 @@ def _ensure_annotation_object(obj):
     if not isinstance(obj, Annotations):
         raise ValueError('Annotations must be an instance of '
                          'mne.Annotations. Got %s.' % obj)
+
+
+@verbose
+def events_from_annotations(raw, event_id=None, regexp=None,
+                            event_id_func=None, verbose=None):
+    """Get events and event_id from an Annotations object.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data for which Annotations are defined.
+    event_id : dict | None
+        Dictionary of string keys and integer values as used in mne.Epochs
+        to map annotation descriptions to integer event codes. Only the
+        keys present will be mapped and the annotations with other descriptions
+        will be ignored. If None, all descriptions of annotations are mapped
+        and assigned arbitrary unique integer values.
+    regexp : str | None
+        Regular expression used to filter the annotations whose
+        descriptions is a match.
+    event_id_func : None | callable
+        What to do for events not found in ``event_id``. Must take one
+        argument and return an ``int``. If None, ``event_id_func`` behaves like
+        a counter that increments every time is called.
```

Comment:
Actually looking at the API, I find this a bit convoluted. Why have both `event_id` and `event_id_func`? It seems cleaner to have the user give just `event_id` dict.

Plus the `event_id` description says that any keys not in `event_id` will be ignored, but this makes it sound like they will not...?

Reply:
... maybe this would be cleared up with a sufficient `Examples` section to show how these things relate to one another.

Reply:
So put a public API in `master` that we plan to walk back before release? This seems like an unusual approach.

I'd prefer to write the public facing API the way we want it, and for each reader, make it work with this API (expanding it only if necessary). It seems like a safer and cleaner approach, even though it's more work in this PR (since it will need to do some conversions using this function to satisfy backward compat).

Reply:
@jona-sassenhagen we'll need your vision here. Why do you think it was necessary to have both event_id and event_id_func in eeglab reader? what is the use case?

Reply:
> Actually looking at the API, I find this a bit convoluted. Why have both `event_id` and `event_id_func`? It seems cleaner to have the user give just `event_id` dict.

The idea of having them both makes sense to me. I'll take the things from A, if they are not there I'll take them from B. 

> Plus the `event_id` description says that any keys not in `event_id` will be ignored, but this makes it sound like they will not...?

I overlooked that point. `Only the keys present will be mapped and the annotations with other descriptions will be ignored.` this is not true. It relegates it to `event_id_func`, you are right.

Maybe `event_id` should be removed in favor of `event_id_func` and have a default behavior that takes an `event_id` and have all the logic inside.

```py
def my_event_id_generator(event_id=None, missing_strategy=None):
     """ bla bla 
         missing_strategy = "drop" | "incremental integer" | callable  
     """
...

def events_from_annotations(raw, regexp=None, event_id_func=None, verbose=None):
...
```
The namings are completly wrong, but just to illustrate. 

```py
    raw = read_raw_fif(fif_fname)
    events = mne.find_events(raw)
    event_id = {
        'Auditory/Left': 1,
        'Visual/Smiley': 32,
        'BlaBla42': 42,
    }
    # just drop them
    events, new_events_id = events_from_annotations(raw, event_id_func=my_event_id_generator(event_id, missing_strategy='drop'
)
    # give them an id
    events, new_events_id = events_from_annotations(raw, event_id_func=my_event_id_generator(event_id, missing_strategy='incremental integer'
)
    # use a callable to take the naming to generate the id
    events, new_events_id = events_from_annotations(raw, event_id_func=my_event_id_generator(event_id, missing_strategy=_strip_integer
)
```

Reply:
> The idea of having them both makes sense to me. I'll take the things from A, if they are not there I'll take them from B.

I don't see the point. It splits what should be one operation over two params.

> Maybe event_id should be removed in favor of event_id_func and have a default behavior that takes an event_id and have all the logic inside.

This is more what I'm thinking -- there should be one way to do it.

For simplicity and consistency, maybe we should call it `event_id`:

- if dict: if it has the key, use the event, if not, discard it
- if callable (takes a key as single input): returns an event ID number to keep the event, return None to discard it

Then for `eeglab` and backward compat we supply the user with some simple wrapper function that does this auto-counting business to combine the two params. It looks like this is what you have in mind with your `my_event_id_generator` function, but no need for `missing_strategy` -- the `drop` condition is handled by just passing `dict` directly as `event_id`; the incremental is what `my_event_id_generator` does; and the `callable` is an advanced thing that a user should really write their own few-line function to handle.